src/components/data_portal/DataPortalShared.tsx

```diff
@@ -210,32 +231,158 @@ export const standaloneStyle = makeStyles((theme: Theme) =>
+        return { ...acc, ...obj }
+      }, {})
+    } else return res.reduce((acc, array) => acc.concat(array), [])
+  } else {
+    return [id_set]
+  }
+}
+
+//given a valid researcher, study, or participant id,
+//an array of study ids is returned
+export async function generate_study_ids(id_set) {
+  if (typeof id_set === "string") {
+    let { data: parents } = await LAMP.Type.parent(id_set)
+    //if study exists, return study, this is a participant
+    if (parents?.Study) return [parents.Study]
+    //else if researcher exists
+    //this is a study, so we return id_set
+    else if (parents?.Researcher) {
+      return [id_set]
+    }
+    //if nothing exists, this is a researcher, so we get the list of studies
+    //under their purview and return that
+    else {
+      let res = await LAMP.Study.allByResearcher(id_set)
+      return res.map((study) => study.id)
+    }
+  } else if (Array.isArray(id_set)) {
+    const res = await Promise.all(id_set.map((id) => generate_study_ids(id)))
     //now, res is an array of arrays. let's combine them
     return res.reduce((acc, array) => acc.concat(array), [])
   } else {
     return [id_set]
   }
 }
+
+//given an id or array of ids, returns a dictionary containing
+//key value pairs of activity ids and activity names
+export async function generate_activity_dict(
+  id_set,
+  token,
+  included_details = ["name", "spec"],
+  already_reduced = false
+) {
+  let id_list
+  if (already_reduced) id_list = id_set
+  else id_list = await generate_study_ids(id_set)
+
+  //pull an array of arrays of studies
+  //turns out we don't want to do this with LAMP.Activity.allByStudy
+  //it fetches too much information and is inefficient.
+  //Let's use a jsonata query instead!
+  const res = await Promise.all(
+    id_list.map((id) =>
+      jsonataFetch(queryDictionary["activityFromStudy"](id), token.username, token.password, token.server)
+    )
+  )
+  //flatten the array of arrays
+  let studyArray = res
+    .filter((elem) => Array.isArray(elem))
+    .reduce((acc, array) => (acc as Array<any>).concat(array), [])
+  //return a dictionary with activity ids and names
+  return (studyArray as Array<any>).reduce((acc, example) => {
+    let result_obj = {}
+    included_details.forEach((key) => {
+      if (example[key]) result_obj[key] = example[key]
+    })
+    return { ...acc, ...{ [example["id"]]: result_obj } }
+  }, {})
+}
+
+//this function takes a LAMP timestamp and returns
+//a nicely formatted UTC string
+export function ts_to_UTC_String(timestamp) {
+  let date = new Date(timestamp)
+  return date.toLocaleString()
+}
+
+//this function takes an object and returns an object
+//with the same keys where the values have been stringified
+export function stringify_obj_values(obj, omit_on_return: Array<string> = []) {
+  let res = {}
+  Object.keys(obj).forEach((key) => (res[key] = JSON.stringify(obj[key])))
+  omit_on_return.forEach((key) => delete res[key])
```

Comment:
Alternatively, `Object.toEntries` converts `{ abc: 123, ... }` into `[ [ abc, 123], ... ]`. Use that and `Object.fromEntries`, which allow you to use `.map()` and `.filter()` in a more idiomatic way. (Don't mix functional syntax and the `delete` keyword/loops/etc.)